orttraining/orttraining/training_ops/cuda/collective/nccl_kernels.cc

```diff
@@ -43,78 +53,201 @@ Status NcclAllGather::ComputeInternal(OpKernelContext* context) const {
   const size_t element_size = onnx_type->Size();
   ncclDataType_t dtype = GetNcclDataType(onnx_type);
 
-  // Count total number of elements to AllGather.
-  int64_t total_count = 0;
-  for (int i = 0; i < context->InputCount(); i++) {
-    const Tensor* input_tensor = context->Input<Tensor>(i);
-    total_count += input_tensor->Shape().Size();
-  }
-
   // AllGather requires every rank to receive the same amount of data, and
   // slows down significantly if the data is not aligned.  Nvidia recommends 32-byte alignment,
   // so pad to multiple of 32 and world size.
   // Note: the alignment here needs to be kept in-sync with the alignment in zero_optimizer_graph_builder.cc
-  const int64_t alignment = size * 32;
-  const int64_t padded_count = total_count + alignment - (total_count % alignment);
-  const int64_t padded_size = padded_count * element_size;
-  auto fusion_buffer = GetScratchBuffer<void>(padded_size);
-  void* fusion_data = fusion_buffer.get();
+  if (partition_even_) {
+    ORT_ENFORCE(num_input_readies_ == 0);
+    // Count total number of elements to AllGather.
+    int64_t total_count = 0;
+    for (int i = 0; i < context->InputCount(); i++) {
+      const Tensor* input_tensor = context->Input<Tensor>(i);
+      total_count += input_tensor->Shape().Size();
+    }
+    const int64_t alignment = size * 32;
+    const int64_t padded_count = total_count + alignment - (total_count % alignment);
+    const int64_t padded_size = padded_count * element_size;
+    auto fusion_buffer = GetScratchBuffer<void>(padded_size);
+    void* fusion_data = fusion_buffer.get();
+
+    // Calculate the range of inputs this rank will send.
+    ORT_ENFORCE(padded_count % size == 0);
+    const int64_t rank_count = padded_count / size;
+    const int64_t rank_bytes = rank_count * element_size;
+    const int64_t rank_start = rank * rank_bytes;
+    const int64_t rank_end = rank_start + rank_bytes;
+
+    // Copy this rank's inputs to fusion buffer.
+    int64_t offset = 0;
+    for (int i = 0; i < context->InputCount(); i++) {
+      const Tensor* input_tensor = context->Input<Tensor>(i);
+      const int64_t tensor_bytes = input_tensor->SizeInBytes();
+
+      // Only copy inputs this rank needs to send.
+      if (rank_start <= offset && offset < rank_end) {
```

Comment:
I think this is correct with <= and <, but just confirming given that the are inclusive start...end ranges elsewhere.